server/src/main/java/org/opensearch/common/xcontent/JsonToStringXContentParser.java

```diff
@@ -72,21 +73,31 @@ public XContentParser parseObject() throws IOException {
         builder.startObject();
         LinkedList<String> path = new LinkedList<>(Collections.singleton(fieldTypeName));
         while (currentToken() != Token.END_OBJECT) {
-            parseToken(path);
+            parseToken(path, null);
         }
-        builder.field(this.fieldTypeName, keyList);
-        builder.field(this.fieldTypeName + VALUE_SUFFIX, valueList);
-        builder.field(this.fieldTypeName + VALUE_AND_PATH_SUFFIX, valueAndPathList);
+        // deduplication the fieldName,valueList,valueAndPathList
+        builder.field(this.fieldTypeName, new HashSet<>(keyList));
+        builder.field(this.fieldTypeName + VALUE_SUFFIX, new HashSet<>(valueList));
+        builder.field(this.fieldTypeName + VALUE_AND_PATH_SUFFIX, new HashSet<>(valueAndPathList));
         builder.endObject();
         String jString = XContentHelper.convertToJson(BytesReference.bytes(builder), false, MediaTypeRegistry.JSON);
         return JsonXContent.jsonXContent.createParser(this.xContentRegistry, this.deprecationHandler, String.valueOf(jString));
     }
 
-    private void parseToken(Deque<String> path) throws IOException {
+    /**
+     * @return true if the child object contains no_null value, false otherwise
+     */
+    private boolean parseToken(Deque<String> path, String currentFieldName) throws IOException {
```

Comment:
Do we need to pass `currentFieldName` down? 

I see it getting overwritten on line 97, but the value passed is never read anywhere, from what I can see.

Reply:
Yeah -- I was able to remove it from the parameter list without breaking anything: https://github.com/opensearch-project/OpenSearch/compare/main...msfroh:OpenSearch:fix_null_flat_object#diff-9b94311ef376a05bce2bbe32e99582f739378209a89aaa6dc4bc192e4b8f2660R90